python/tchannel/tornado/connection.py

```diff
@@ -222,6 +223,17 @@ def _loop(self):
                     future = self._outstanding.get(context.message_id)
                 else:
                     future = self._outstanding.pop(context.message_id)
+
+                    if context.message.message_type == Types.ERROR:
+                        protocol_error = (
+                            self.response_message_factory.build_protocol_error(
+                                context.message,
+                                context.message_id,
+                            )
+                        )
+                        future.set_result(protocol_error)
+                        continue
```

Comment:
This is because we will mutate the future object for tracing staff after we receive it but we can't guarantee the execution order of this add_done_callback, and in order to send same future object to  whoever yield this future, we create two futures.